tidy3d/components/base.py

```diff
@@ -517,6 +524,9 @@ def __ge__(self, other):
         """define >= for getting unique indices based on hash."""
         return hash(self) >= hash(other)
 
+    def __eq__(self, other):
```

Comment:
Currently `pydantic.BaseModel` uses a dict conversion, and it is already a [known bottleneck](https://github.com/pydantic/pydantic/issues/1494). Using a json string is more constly both due to conversion and comparison, so I'd vote not to use this.
Is there any particular reason to compare this way instead of with dicts?

Reply:
A few things.
1. the equality checking is mainly just used in tests, eg to ensure that a sim loaded from file equals the original. However, I agree we should try to use a more performant solution in case people are using this more.
2. supporting `ArrayLike` with `np.ndarray` means that simulations are no longer hashable because `np.ndarray` is not hashable. without defining my own `__eq__` for base, it was using dict comprehension and I was getting errors about `np.ndarray` not being hashable. The (hacky) fix was to convert this to json string for comparison.

I don't really see too many alternatives to this except for.
1. reverting to not store ArrayLike as `np.ndarray`, to retain hashability.
2. subclassing `np.ndarray` and providing `hash` and `eq` methods & using pydantic's dict comprehension.
3. writing our own dict comprehension that does equality checking differently if the compared values are `np.ndarray` (eg using `np.all`) 

But I'm not sure we can get around a dict comprehension here. Do you have any ideas?

Reply:
I think number 3 might be the best option here, then. If that's going to be too complicated, then maybe leave it as is with a TODO note, since it is only for testing at this point.

Reply:
I took a stab at subclassing [here](https://github.com/flexcompute/tidy3d/blob/84ba787958c535a16118e9b48aceda41d1cb13a2/tidy3d/components/types.py#L30-L44) and [here](https://github.com/flexcompute/tidy3d/blob/84ba787958c535a16118e9b48aceda41d1cb13a2/tidy3d/components/types.py#L76-L80). and It's a bit messy, but seems to work for hashing. not sure If this is a very good solution or we should rethink things..

Reply:
Hard to say. Numpy arrays are much easier to work with and avoid explicit loops. The drawback of sub-classing is that operation results would be ndarray, not TidyNDArray, right? But then, again the validators in ArrayLike woud convert anything the requires hashing to TidyNDArray, so that's not too bad I guess.